stake-pool/program/src/state.rs

```diff
@@ -115,3 +118,176 @@ impl State {
         }
     }
 }
+
+const MAX_VALIDATOR_STAKE_ACCOUNTS: usize = 1000;
+
+/// Storage list for all validator stake accounts in the pool.
+#[repr(C)]
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct ValidatorStakeList {
+    /// False if not yet initialized
+    pub is_initialized: bool,
+    /// List of all validator stake accounts and their info
+    pub validators: Vec<ValidatorStakeInfo>,
+}
+
+/// Information about the singe validator stake account
+#[repr(C)]
+#[derive(Clone, Copy, Debug, Default, PartialEq)]
+pub struct ValidatorStakeInfo {
+    /// Account pubkey
+    pub account: Pubkey,
```

Comment:
Just to double check, this is a validator pubkey used to generate a program address, correct?

Reply:
Hmm... The plan was for it to be a stake account voting for a particular validator. All deposits from the users will be merged into one of these accounts.

Reply:
Ok good, that's the right track.  This pubkey then should be the validator pubkey that the stake account is voting for, and the actual address of the stake account that everyone merges into is generated using `Pubkey::find_program_address(&[validator_pubkey, stake_pool_pubkey], STAKE_POOL_PROGRAM_ID)`.  Does that make sense?

Reply:
Not yet :) Can you please explain why? As I know stake account does not sign anything and there is no need to make it a program address. The important things are its Withdrawer/Staker authorities and we will set them to the program addresses of the stake pool.

Reply:
You're right that since we have the withdraw authority in the program, we don't need to sign for the stake account.

It'll be easier to get the stake account associated with a validator using a `find_program_address` rather than loading and deserializing all 1000 accounts to see where they're delegated.  For example, any time we add a new validator to stake to, how do we check that the validator doesn't already have a stake account in the pool?  Alternatively, we could additionally store the pubkey of the validator in the `ValidatorStakeInfo` and then use that.

A program address will also make things easier when warming up new stakes to merge into the stake pool from the stake-o-matic.  If we have a convention that stakes warming up will be stored at `find_program_address([validator_pubkey, stake_pool_pubkey, 'warming'], stake_pool_program_id)` we can easily find and merge accounts for deposits.  Otherwise, we need to get all stake accounts controlled by the staker and then match them correctly each time.

Eventually, we'll want to allow people to deposit `cold` stakes that aren't assigned to any validators, so some conventions about how to manage those will make that integration simpler in the long run.

Anyway, that's a bunch of ideas, but I think it will simplify the stake-o-matic integration and further development.

Reply:
Right, this is a clever approach to finding stake address for a particular validator.

Will this program address generation be faster than doing a data filtering (like we do when enumerating stake accounts of a particular stake pool in the current CLI)?

Regarding a different set of addresses for warming up, is it something for the future? Like another set of addresses for warming stake accounts which will be periodically merged to the active stake accounts managed by the pool?

Reply:
Yes, it will absolutely be faster -- taking in 1000 accounts and deserializing all of them in one program instruction will probably  be impossible with the current compute restrictions, but we can probably do 30-50 calls to `find_program_address`.

And that's exactly right, for v2 we'll want to consider deposits of non-warmed up stakes.  Even before that though, as you mention, the stake pool manager will need to manage warming up and deactivating stakes for 1000 validators through stake-o-matic (or another system if they prefer), so having deterministic addresses will make the stake-o-matic much simpler too.

Reply:
Hm, I do not think we need to do it on-chain. If we filter accounts off-chain and simply pass the accounts needed to the program it will just need a single call to verify validity of the data.
But still there is a lot of querying and filtering if we use blunt force to find accounts even off-chain. My idea was to save 2 pubkeys (validator's and stake account) for each stake-pool-managed account, but this will make each record 60% larger. So I guess generating program address for it is fine. Do we need to save nonce or bump seeds so it gets properly deterministic and faster? Or just use `find_program_address`?

Reply:
Also I had a question about adding stake accounts to the pool. So basically right now we need it to be with 0 balance, because otherwise it will screw up the pool tokens calculation (like there will be 10 SOL in the pool even before first tokens are minted). But can account be voting for someone and be pre-warmed if the balance is 0 SOL? And will the account be actually dead if we leave it with 0 SOL until the next rent payment?

My idea was to do the following inside the single transaction:
1. Create new stake account (using program address derived from the validator pubkey), give it 0 SOL
2. Join it to the pool, at this stage stake pool knows the validator this account is planning to vote for even though staking itself was not done yet.
3. Deposit to this account through a regular `deposit` command, get tokens in return
3.1. We'll need a special case in `deposit` instruction which will check if stake account is already staking and if not — do the staking on deposit.

Is it how you see it as well or is there another way?

Reply:
> Hm, I do not think we need to do it on-chain. If we filter accounts off-chain and simply pass the accounts needed to the program it will just need a single call to verify validity of the data.

I'm not sure what "it" is referring to here.  I'm thinking specifically of one example: adding a new validator account.  To add a new account following the model from your document, we need a way to know that an account doesn't already exist for a validator, on-chain.  If we only store the validator address and use `find_program_address` for the stake accounts, then we can easily do this.

> My idea was to save 2 pubkeys (validator's and stake account) for each stake-pool-managed account, but this will make each record 60% larger. So I guess generating program address for it is fine.

You can go that route too if you prefer, I can leave that up to you :-)

> Do we need to save nonce or bump seeds so it gets properly deterministic and faster? Or just use `find_program_address`?

You might have to tune this depending on how many BPF instructions the program is using.  We were thinking it would be easier to just `find_program_address` each time as a standard, but if you're running into other issues, you might need to also store the bump seed.

> Also I had a question about adding stake accounts to the pool.

I see what you mean -- for `Join / Remove Validator-Associated Stake Account to the Pool`, you're wondering about existing balance, correct?  The first solution to me would be to have these instructions work like deposit / withdraw -- you require the account to have a fully active stake already and just mint or burn pool tokens as if it were a deposit / withdrawal.  You're right, the current stake program does not allow merging a stake with 0 active delegation to a stake with some active delegation.  Alternatively, you can require these accounts to be rent-exempt and have 1 lamport of active delegation and do the sequence of instructions that you describe.

Reply:
Thanks, you convinced me, I'm going `find_program_address` route :)

Regarding the last problem with adding account to the pool I think your idea of minting/burning token will be the easiest to implement. So in fact the new `join` instruction will be like the old `deposit` and the new `leave` like the old `claim`.

Reply:
Sounds great!